src/group_theory/iwasawa.lean

```diff
@@ -0,0 +1,655 @@
+
+**Theorem (Iwasawa criterion).** Let G be a group with a primitive action on a set X.
+Assume given an Iwasawa structure, namely, 
+for all x:X, an abelian subgroup T x of G such that 
+1. ∀ g:G x:X, T (g x) = g (T x) g⁻¹ 
+2. The union of all T x generates G
+Then every normal subgroup of G that acts nontrivially on X contains
+the derived subgroup D(G) of G.
+
+**Corollary.** Let G be a perfect group (D(G)=G) 
+with a faithful and primitive action on a set X.
+Assume given an Iwasawa structure.
+Then G is simple.
+
+## Simplicity of special linear groups
+
+**Theorem.** Let K be a field and let n be an integer such that n ≥ 2.
+Unless n = 2 and K has ≤ 3 elements, the group PSL(n,K) is simple.
+
+
+## Simplicity of alternate groups
+
+**Theorem.** Let n be an integer such that n ≥ 5. 
+The alternate group A(n) is simple.
+
+-/
+
+
+section Maximal_Subgroups
+variables {G : Type*} [group G]
+
+/-- An subgroup is maximal if it is maximal in the collection of proper subgroups. -/
+class subgroup.is_maximal (K : subgroup G) : Prop := 
+  (out : is_coatom K)  
+
+theorem is_maximal_def {K : subgroup G} : K.is_maximal ↔ is_coatom K := ⟨λ h, h.1, λ h, ⟨h⟩⟩
+
+theorem is_maximal.ne_top {K : subgroup G} (h : K.is_maximal) : K ≠ ⊤ := (is_maximal_def.1 h).1
+
+theorem is_maximal_iff {K: subgroup G} : K.is_maximal ↔
+  K ≠ ⊤ ∧ ∀ (H: subgroup G) g, K ≤ H → g ∉ K → g ∈ H → H = ⊤ :=
+begin
+  split, 
+  
+  intro hK, split, exact is_maximal.ne_top hK,
+  intros H g hKH hgK hgH,
+  suffices : K < H,
+    apply (is_maximal_def.1 hK).2, assumption,
+  have zKH : K ≠ H := 
+  begin 
+    simp, intro z, rw z at hgK, exact hgK hgH,
+  end,
+  exact (ne.le_iff_lt zKH).mp hKH,
+
+  rintros ⟨ hG, hmax ⟩,
+  split, split, assumption,
+  introsI H hKH,
+  obtain ⟨ g, hgH, hgK ⟩ := (set.exists_of_ssubset hKH),
+  exact hmax H g (le_of_lt hKH) hgK hgH, 
+end
+
+theorem is_maximal.eq_of_le {K H: subgroup G}
+  (hK : K.is_maximal) (hH : H ≠ ⊤) (KH : K ≤ H) : K = H :=
+eq_iff_le_not_lt.2 ⟨KH, λ h, hH (hK.1.2 _ h)⟩
+
+end Maximal_Subgroups
+
+section Basic_Lemmas
+/-- If G and H have multiplications *
+and φ : G → H is a surjective multiplicative map,
+and if G is commutative, then H is commutative. 
+
+Since I only use it with groups, 
+I should probably use function.surjective.comm_semigroup 
+--/
+lemma surj_to_comm {G H : Type*} [has_mul G] [has_mul H] (φ: mul_hom G H) :
+   function.surjective φ → is_commutative G (*) → is_commutative H (*) :=
+begin
+  intros is_surj is_comm, 
+  apply is_commutative.mk,
+  intros a b,
+    obtain ⟨a', ha'⟩ := is_surj a, obtain ⟨b', hb'⟩ := is_surj b,
+    rw ← ha', rw ← hb', 
+    let z := ⇑φ, let z₂ := φ.to_fun, have : z = z₂  , by refl,
+    rw ← mul_hom.map_mul φ a' b',
+    rw ← mul_hom.map_mul φ b' a',
+    apply φ.congr_arg,
+    refine is_commutative.cases_on is_comm _, intro,
+    exact comm a' b',
+end
+
+/-- f(f⁻¹(t)) = f.range ∩ t  -/
+lemma map_comap_of {α β : Type*} (f: α → β) (t: set β):
```

Comment:
I think the more predictable name would be to use `image_preimage_eq`. Also, these lemmas should move to some basic file about `set`.